domain/upgrade/state/state.go

```diff
@@ -0,0 +1,311 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/collections/set"
+	"github.com/juju/collections/transform"
+	"github.com/juju/errors"
+	"github.com/juju/utils/v3"
+	"github.com/juju/version/v2"
+
+	"github.com/juju/juju/domain"
+)
+
+// State is used to access the database.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState creates a state to access the database.
+func NewState(factory domain.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+func (st *State) getUpgrade(ctx context.Context, tx *sqlair.TX, upgradeUUID string) (info, error) {
+	q := `
+SELECT (uuid, previous_version, target_version, created_at, started_at, created_at) AS &info.* FROM upgrade_info`
```

Comment:
can't you do `SELECT * AS &info.* FROM upgrade_info`? I don't know if sqlair allows this, but it's more readable if it does

Reply:
sqlair does allow this it seems. Fixed